### ISSUE TITLE
[bug][fsdp] FSDP2 manual offload/load corrupts DTensor unshard state under `colocate_all=True`, causing silent non-deterministic logit errors

### ISSUE THREAD

**Issue by Aression** [NovaSky-AI/SkyRL#2215]
## Summary

FSDP2 (composable `fully_shard`) with `colocate_all=True` + `cpu_offload=False` (the default, i.e. `manual_offload=True`) silently produces stale/corrupted logits on a small fraction of tokens after `save_weights_for_sampler`, with no crash or warning. The root cause is the manual offload/load path using a bare `model.to("cpu")` / `model.to(device)`, which bypasses FSDP2's `unshard`/`reshard` state management and corrupts the DTensor shard view. This manifests as non-deterministic logit errors on specific positions, an inflated `max|log_ratio|` that never converges, and occasional `grad_norm` spikes.

## Environment

- SkyRL: `main` @ `53b11553` (also reproduced on `34f6da9d`)
- PyTorch: `2.11.0+cu129` (FSDP2 composable `fully_shard`)
- vLLM: `0.20.2`, transformers: `4.57.6`, peft: `0.18.1`, ray: `2.51.1`
- LoRA rank=64/alpha=64/all-linear on a 4B-param causal LM
- `cpu_offload=False` (default), `colocate_all=True`, `reshard_after_forward=True`; student FSDP across 3 GPUs + 3 colocated vLLM engines (`gpu_memory_utilization=0.45`)

## Reproduction

Config that triggers it (LoRA + colocated engines + default `cpu_offload`):

```yaml
trainer:
  strategy: fsdp
  placement:
    colocate_all: true
  policy:
    fsdp_config:
      cpu_offload: false          # default -> manual_offload = True
      reshard_after_forward: true
    model:
      lora:
        rank: 64
        alpha: 64
generator:
  inference_engine:
    num_engines: 3
    gpu_memory_utilization: 0.45
```

After `init_model`, an FSDP forward on the same probe produces logits matching a CPU reference (argmax all-match, diff < 0.01). After the first `save_weights_for_sampler` — which triggers the offload→load cycle (`_finish_weight_sync` → `offload_to_cpu` → `model.to("cpu")`, then next forward → `_ensure_on_gpu` → `backload_to_gpu` → `model.to(device)`) — the same probe's logits diverge **non-deterministically**:

| run | fsdp vs ref absmean | fsdp vs ref absmax | notes |
|---|---|---|---|
| 1 | 0.79 | 5.97 | wrong at pos 6/9/14 |
| 2 | 0.64 | 9.00 | wrong at pos 6 |
| 3 | 0.81 | 10.88 | wrong at pos 6/14 |
| 4 | 0.03 | 0.49 | correct (1/7) |
| 5 | 0.70 | 5.17 | wrong at pos 6/9/14 |
| 7 | 0.64 | 9.00 | wrong at pos 6 |

Key observations:
- The error only appears **after** the offload→load cycle; init-time forward is correct.
- `absmax ≈ 28.88`, ~4× larger than the bf16/autocast precision ceiling (~7.3) measured on CPU — not a precision error.
- **Non-deterministic**: 6/7 runs wrong, error values vary run-to-run, error positions are fixed but error magnitudes are random.
- Within a single affected forward, an explicit `lm_head.weight.full_tensor()` gather still returns the **correct** weight (diff=0 vs disk), but the forward's logits are wrong (absmax≈23.88) — the forward consumed a stale/corrupted unshard view while post-hoc gather recovers correct params.
- Inserting an extra `torch.save` (a synchronization point) can "accidentally" make a run correct → **timing-sensitive**.
- Not reproducible on CPU (single-process, no distributed shard/all-gather).

## Root cause

`cpu_offload=False` (default) → `manual_offload = not cpu_offload = True`:

https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/backends/skyrl_train/distributed/fsdp_strategy.py#L87

The colocate path then offloads/loads via bare `.to()`, which does not go through FSDP2's `unshard()`/`reshard()`:

https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/backends/skyrl_train/distributed/fsdp_utils.py#L66-L78

https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/backends/skyrl_train/distributed/fsdp_utils.py#L81-L84

Called from the weight-sync / forward cycle:
- `save_weights_for_sampler` → `_finish_weight_sync` → `_offload(policy, offload_model=True)` → `offload_to_cpu` → `offload_fsdp2_model_to_cpu`
- next forward → `_ensure_on_gpu` → `backload_to_gpu` → `load_fsdp2_model_to_gpu`

The bare `model.to(...)` moves DTensor parameters directly without invalidating `_unsharded_param` cached state. After the offload→load cycle, subsequent forwards intermittently consume a stale shard view, producing the non-deterministic logits above.

PyTorch's `FSDPModule._apply` does attempt to guard `.to()` via `reshard()` + `reset_sharded_param()`, but the protection has gaps this path hits:
- `reshard()` is **not recursive** ("This method is *not* recursive"), while `model.to()` is recursive over submodules with their own FSDP state.
- `_unsharded_param` is a lazily-allocated cache that `.to()` does not invalidate, so a forward can read a full tensor pointing at moved/stale storage.

> Note: `fsdp2_load_full_state_dict` has a comment acknowledging "the offload path is still broken for FSDP2", but that only covers the **init** path (`cpu_offload` defaults to `None` at the call site, so the `if cpu_offload:` branch is unreachable under the default config). The **training-time** manual offload/load in `offload_fsdp2_model_to_cpu`/`load_fsdp2_model_to_gpu` has no such guard at all.

## Impact

- `max|log_ratio|` (`current_logprob − old_logprob`) stays at ~34–38 across the full run and is **non-monotonic** (drops to ~32 then rebounds to ~35), never reaching the <5 target — because every step's weight sync re-corrupts the state to a different degree.
- Occasional `grad_norm` spikes (e.g. 462 at one step, 25/20 at others) from the corrupted ratio.
- The error is **silent** (no crash, no warning), and post-hoc `full_tensor()` gathers hide it, making it very hard to diagnose. `CPUOffloadPolicy`'s own docstring states all-gathered params are "freed according to `reshard_after_forward`" — i.e. the native path coordinates with the state machine; the manual `.to()` path does not.

## Expected behavior

The manual offload/load path (or at least the `cpu_offload=False` default under `colocate_all=True`) should either:
1. Go through `FSDPModule.unshard()`/`reshard()` explicitly instead of bare `model.to(...)`, or
2. Invalidate `_unsharded_param` caches on offload/load, or
3. Fall back to the native `CPUOffloadPolicy` (set `cpu_offload=True`), or
4. At minimum, raise/warn when `colocate_all=True` + `cpu_offload=False` + FSDP2 is selected, since this combination silently corrupts.

## Workarounds

- Setting `fsdp_config.cpu_offload: true` avoids the manual `.to()` path (native offload) and eliminates the symptoms in our runs.
- vLLM prefill rescoring (making `current`/`old` come from the same engine) sidesteps the FSDP-side recomputation entirely.

## Related

- #2189 makes `GPUState` explicit (fixes a different redundant-backload bookkeeping bug under profiler; does not address the FSDP2 state corruption here).
- #2206 (open) proposes improving offload/backload semantics for LoRA + `colocate_all=True`, but targets trainable/non-trainable bookkeeping, not the FSDP2 `unshard` state corruption.
- #2062 / #2064 keep LoRA adapter buffers GPU-resident to bypass offload on the **Megatron** backend; the FSDP backend path is unchanged.

Happy to provide a minimal standalone repro if useful. We confirmed the mechanism via same-batch three-way comparison (vLLM decode vs FSDP prefill vs CPU reference) and 7× independent probe runs.

**Comment by SumanthRH** [comment]
@Aression 

Could you provide a minimal reproduction script for the issue you're seeing? Can be a standalone script with some dummy data or a SkyRL training configuration with some public dataset.  Thanks!
